### DISCUSSION TITLE
peer_discovery_backend requires module name specified (at least for Consul)

### DISCUSSION THREAD

**Discussion by tj90241** [rabbitmq/rabbitmq-server#6754] [Questions] [answered]
Compiled 3.10.12 from source and configured it for use with the Consul discovery plugin per documentation, which states:

"To use Consul for peer discovery, set the cluster_formation.peer_discovery_backend to consul or its module name, rabbit_peer_discovery_consul"

So, this should work:
```
{rabbit, [
  {cluster_formation, [
    {peer_discovery_backend, consul},
    {consul, [
...
```

However, when booting `rabbitmq-server`, it failed with:
```
BOOT FAILED
===========
Exception during startup:

error:undef

    consul:supports_registration/0
        args: []
    rabbit_peer_discovery:maybe_register/0, line 124
    rabbit_mnesia:init/0, line 82
    rabbit_boot_steps:-run_step/2-lc$^0/1-0-/2, line 41
    rabbit_boot_steps:run_step/2, line 48
    rabbit_boot_steps:-run_boot_steps/1-lc$^0/1-0-/1, line 19
    rabbit_boot_steps:run_boot_steps/1, line 20
    rabbit:start/2, line 906

Kernel pid terminated (application_controller) ({application_start_failure,rabbit,{undef,{rabbit,start,[normal,[]]}}})

Crash dump is being written to: /var/log/rabbitmq/erl_crash.dump...done
```

So clearly, it can't find the plugin. Which led me to make no other changes to my configuration, other than changing `peer_discovery_backend` to the module name (`rabbitmq_peer_discovery_consul`). This worked, and there was no more `error:undef` thrown for `consul:supports_registration/0`.

**Comment by tj90241** [comment]
The configuration schema also seems to be namespaced differently?

I'm groking: https://www.rabbitmq.com/cluster-formation.html

which says e.g.:
```
cluster_formation.peer_discovery_backend = consul

cluster_formation.consul.host = consul.eng.example.local
cluster_formation.consul.acl_token = acl-token-value
```

However, I had to namespace everything like this based on the source code, which worked:
```
{rabbit, [
  {cluster_formation, [
    {peer_discovery_backend, rabbitmq_peer_discovery_consul},
    {peer_discovery_consul, [
      {consul_host, "consul.service.consul.<REDACTED>"},
      {consul_acl_token, "<REDACTED>"},
      {consul_port, 8501},
      {consul_scheme, https},
      {consul_ssl_options, [
        {cacertfile, "/etc/rabbitmq/ssl/consul_chain.pem"},
        {certfile,   "/etc/rabbitmq/ssl/consul_cert.pem"},
        {keyfile,    "/etc/rabbitmq/ssl/consul_privkey.pem"}
      ]},
      {consul_svc, rabbitmq},
      {consul_svc_addr_auto, false},
      {consul_svc_addr, "<REDACTED>"},
      {consul_use_longname, true}
```

except `consul_ssl_options` -- those don't seem to be respected, anymore after having been added in https://github.com/rabbitmq/rabbitmq-server/commit/8f8a138e0d371ce882e917a54608fc99bbfaa0bc

**Comment by lukebakken** [comment, marked as answer]
The configuration you have provided, that begins with `{rabbit, [`, is in **Erlang term format**, and goes into the `advanced.config` file.

In general, the documentation refers to the **`ini`-style** configuration that should be in the `rabbitmq.conf` file. You should be able to provide all of your Consul configuration in that file and _not_ use the Erlang term format configuration.

https://www.rabbitmq.com/cluster-formation.html#peer-discovery-consul

So, in your `/etc/rabbitmq/rabbitmq.conf` file, have the following contents:

```
cluster_formation.peer_discovery_backend = consul

cluster_formation.consul.host = consul.eng.example.local
cluster_formation.consul.acl_token = acl-token-value
cluster_formation.consul.port = 8501
cluster_formation.consul.scheme = https
cluster_formation.consul.ssl_options.cacertfile = /etc/rabbitmq/ssl/consul_chain.pem
cluster_formation.consul.ssl_options.certfile = /etc/rabbitmq/ssl/consul_cert.pem
cluster_formation.consul.ssl_options.keyfile = /etc/rabbitmq/ssl/consul_privkey.pem
cluster_formation.consul.svc = rabbitmq
cluster_formation.consul.svc_addr_auto = false
cluster_formation.consul.svc_addr = <REDACTED>
cluster_formation.consul.use_longname = true
```

**Reply by tj90241** [reply]
hmm, thanks -- sorry for wasting your time!

**Reply by lukebakken** [reply]
No worries, let us know if you have issues with the configuration I provided. The fact you can configure RabbitMQ two different ways is not amenable to new users.
